### PR TITLE
update py-nbconvert

### PR DESCRIPTION
**Pull request by healther**
* add setuptools dependency, like all the other jupyter packages
  it seems to be using setuptools for some commands all the time
  but requires it for the newest version
* added dependencies, not necessarily only needed for the latest one
* depends on new packages (defusedxml, pandocfilters, testpath)

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-nbconvert/package.py [hunk 1]
```diff
@@ -13,21 +13,33 @@ class PyNbconvert(PythonPackage):
...
     url      = "https://github.com/jupyter/nbconvert/archive/4.2.0.tar.gz"
 
+    version('5.6.0', sha256='21a69fe5756a3af8e76dcc105217cc3bea4d8f23dbff5727d9fd1e4388b53beb')
     version('4.2.0', sha256='32394be5a20f39f99fabfb9b2f2625df17f1c2a6699182ca41598e98e7cc9610')
     version('4.1.0', sha256='459f23381411fd1ff9ec5ed71fcd56b8c080d97b3a1e47dae1c5c391f9a47266')
     version('4.0.0', sha256='00e25eeca90523ba6b774b289073631ef5ac65bb2de9774e9b7f29604516265c')
 
     depends_on('py-pycurl', type='build')
+    depends_on('py-setuptools', type='build',                     when='@5:')
     depends_on('python@2.7:2.8,3.3:')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-nbconvert/package.py]:
```suggestion
    depends_on('python@2.7:2.8,3.5:', type=('build', 'run'))
```

File: var/spack/repos/builtin/packages/py-nbconvert/package.py [hunk 2]
```diff
@@ -13,21 +13,33 @@ class PyNbconvert(PythonPackage):
...
+    depends_on('py-jinja2@2.4:', type=('build', 'run'),           when='@5:')
     depends_on('py-pygments', type=('build', 'run'))
     depends_on('py-traitlets', type=('build', 'run'))
+    depends_on('py-traitlets@4.2:', type=('build', 'run'),        when='@5:')
     depends_on('py-jupyter-core', type=('build', 'run'))
     depends_on('py-nbformat', type=('build', 'run'))
+    depends_on('py-nbformat@4.4.0:', type=('build', 'run'),       when='@5:')
     depends_on('py-entrypoints', type=('build', 'run'))
+    depends_on('py-entrypoints@0.2.2:', type=('build', 'run'),    when='@5:')
     depends_on('py-tornado', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-nbconvert/package.py]: This dependency is optional, I would either remove it or add a variant for `+serve`.

File: var/spack/repos/builtin/packages/py-nbconvert/package.py [hunk 3]
```diff
@@ -13,21 +13,33 @@ class PyNbconvert(PythonPackage):
...
     depends_on('py-jupyter-client', type=('build', 'run'))
+    depends_on('py-jupyter-client@5.3.1:', type=('build', 'run'), when='@5:')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-nbconvert/package.py]: This dependency is optional, I would either remove it or add a `+execute` variant.

File: var/spack/repos/builtin/packages/py-nbconvert/package.py [hunk 4]
```diff
@@ -13,21 +13,33 @@ class PyNbconvert(PythonPackage):
...
     depends_on('py-jinja2', type=('build', 'run'))
+    depends_on('py-jinja2@2.4:', type=('build', 'run'),           when='@5:')
     depends_on('py-pygments', type=('build', 'run'))
     depends_on('py-traitlets', type=('build', 'run'))
+    depends_on('py-traitlets@4.2:', type=('build', 'run'),        when='@5:')
     depends_on('py-jupyter-core', type=('build', 'run'))
     depends_on('py-nbformat', type=('build', 'run'))
+    depends_on('py-nbformat@4.4.0:', type=('build', 'run'),       when='@5:')
     depends_on('py-entrypoints', type=('build', 'run'))
+    depends_on('py-entrypoints@0.2.2:', type=('build', 'run'),    when='@5:')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-nbconvert/package.py]: Did none of these dependencies have a version requirement in 4.X? The developers may have not known them. I think it's safe to require these versions for every version, just to simplify the package.

**Comment by adamjstewart** [comment]
Ping @healther

**Comment by healther** [comment]
I'd prefer not to change up any old versions (at least not without checking the versions manually, for which I don't have the time)...

At some point there was some discussion about adding logic to parse `requirements.txt` files, did we ever go anywhere with that? Unfortunately work has been somewhat chaotic the last couple of week

**Comment by healther** [comment]
@adamjstewart: The dependencies are plain, at least up to 4.2.0 (the latest in spack), I'd prefer not to restrict dependencies unless we find an actual breakage, it's much harder to debug a concretisation error than it is to debug a build or run failure

**Comment by adamjstewart** [comment]
That's fine, you still need to address the other comments though.

File: var/spack/repos/builtin/packages/py-nbconvert/package.py [hunk 5]
```diff
@@ -13,21 +13,37 @@ class PyNbconvert(PythonPackage):
     homepage = "https://github.com/jupyter/nbconvert"
     url      = "https://github.com/jupyter/nbconvert/archive/4.2.0.tar.gz"
 
+    version('5.6.0', sha256='21a69fe5756a3af8e76dcc105217cc3bea4d8f23dbff5727d9fd1e4388b53beb')
     version('4.2.0', sha256='32394be5a20f39f99fabfb9b2f2625df17f1c2a6699182ca41598e98e7cc9610')
     version('4.1.0', sha256='459f23381411fd1ff9ec5ed71fcd56b8c080d97b3a1e47dae1c5c391f9a47266')
     version('4.0.0', sha256='00e25eeca90523ba6b774b289073631ef5ac65bb2de9774e9b7f29604516265c')
 
+    variant('execute', default=True)
+    variant('serve', default=True)
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-nbconvert/package.py]: These variants need descriptions

File: var/spack/repos/builtin/packages/py-nbconvert/package.py [hunk 6]
```diff
@@ -13,21 +13,37 @@ class PyNbconvert(PythonPackage):
...
+
     depends_on('py-pycurl', type='build')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-nbconvert/package.py]: This dependency might only be needed for old versions of Python. Can you try:
```suggestion
    depends_on('py-pycurl', when='^python@:2.7.8', type='build')
```
and see if you can build without it on newer versions?

File: var/spack/repos/builtin/packages/py-nbconvert/package.py [hunk 7]
```diff
@@ -13,21 +13,37 @@ class PyNbconvert(PythonPackage):
...
     depends_on('py-jinja2', type=('build', 'run'))
+    depends_on('py-jinja2@2.4:', type=('build', 'run'),           when='@5:')
     depends_on('py-pygments', type=('build', 'run'))
     depends_on('py-traitlets', type=('build', 'run'))
+    depends_on('py-traitlets@4.2:', type=('build', 'run'),        when='@5:')
     depends_on('py-jupyter-core', type=('build', 'run'))
     depends_on('py-nbformat', type=('build', 'run'))
-    depends_on('py-entrypoints', type=('build', 'run'))
+    depends_on('py-nbformat@4.4.0:', type=('build', 'run'),       when='@5:')
+    depends_on('py-entrypoints', type=('build', 'run'),           when='+serve')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-nbconvert/package.py]:
```suggestion
    depends_on('py-entrypoints', type=('build', 'run'))
```

File: var/spack/repos/builtin/packages/py-nbconvert/package.py [hunk 8]
```diff
@@ -13,21 +13,37 @@ class PyNbconvert(PythonPackage):
...
+    depends_on('py-entrypoints@0.2.2:', type=('build', 'run'),    when='@5:+serve')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-nbconvert/package.py]:
```suggestion
    depends_on('py-entrypoints@0.2.2:', type=('build', 'run'))
```

File: var/spack/repos/builtin/packages/py-nbconvert/package.py [hunk 9]
```diff
@@ -13,21 +13,37 @@ class PyNbconvert(PythonPackage):
...
     depends_on('py-tornado', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-nbconvert/package.py]:
```suggestion
    depends_on('py-tornado@4.0:', when='+serve', type=('build', 'run'))
```

File: var/spack/repos/builtin/packages/py-nbconvert/package.py (hunk 6 above)
Comment (healther) [reply var/spack/repos/builtin/packages/py-nbconvert/package.py]: Still builds with python@3.7.4, I think you are correct

**Comment by healther** [comment]
sorry for the delay, I was on vacation the last 3 weeks

File: var/spack/repos/builtin/packages/py-nbconvert/package.py [hunk 10]
```diff
@@ -13,21 +13,36 @@ class PyNbconvert(PythonPackage):
...
-    depends_on('py-pycurl', type='build')
-    depends_on('python@2.7:2.8,3.3:')
+    variant('execute', default=True, description='Include jupyter-client')
+    variant('serve', default=True, description='Include a webserver')
+
+    depends_on('py-pycurl', type='build',                         when='^python@2.7:2.8')
+    depends_on('py-setuptools', type='build',                     when='@5:')
+    depends_on('python@2.7:2.8,3.3:', type=('build', 'run'))
+    depends_on('python@2.7:2.8,3.5:', type=('build', 'run'),      when='@5:')
     depends_on('py-mistune', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-nbconvert/package.py]:
```suggestion
    depends_on('py-mistune@0.8.1:1.999', type=('build', 'run'))
```
At least in the latest version

Comment (healther) [reply var/spack/repos/builtin/packages/py-nbconvert/package.py]: I know, but it works without (at least for me) and we currently don't have a newer mistune version (fixed by #13887). I'm unsure whether to put the version restriction in here now or not

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-nbconvert/package.py]: Just put it in now, then when someone adds a newer version of mistune things don't break

**Pull request merged**